src/common/utils/singleton.ts

```diff
@@ -88,3 +88,94 @@ export class Singleton {
 }
 
 export default Singleton;
+
+interface CreateRequirements<T extends typeof Singleton> {
+  builds: T,
+  args?: ConstructorParameters<T> | (() => ConstructorParameters<T>),
+  requires?: Iterable<typeof Singleton>,
+}
+
+interface DeclareOpts {
+  requires?: Iterable<typeof Singleton>,
+
+}
+
+/**
+ * A builder class for declaring the dependency graph of a set of singletons
+ *
+ * This is useful because it allows the code to declare the requirements
+ */
+export class CreateSingletons {
+  rawRequirements: CreateRequirements<typeof Singleton>[] = [];
```

Comment:
This should probably be marked as private?

Reply:
Yes